### PR TITLE
Expose InsertError.index

### PR DESCRIPTION
**Pull request by zedalaye**
Expose Google::Apis::BigqueryV2::InsertAllTableDataResponse::InsertError.index in Google::Cloud::Bigquery::InsertResponse::InsertError

### CODE REVIEW

**Comment by quartzmo** [comment]
I pulled the branch and ran acceptance tests, which pass. Since there is no acceptance test exercising these errors, I also did some quick manual testing:

```ruby
irb(main):005:0> rows = [{ name: "silvano", breed: "the cat kind",      id: 4, dob: Time.now.utc }, { name: "silvano", breed: "the cat kind",      id: 4, dob: "BAD VALUE" }]
irb(main):006:0> insert_response = table.insert rows
=> #<Google::Cloud::Bigquery::InsertResponse:0x007fcbcaa91260 @rows=[{"name"=>"silvano", "breed"=>"the cat kind", "id"=>4, "dob"=>"2017-07-06 23:00:18.262450+00:00"}, {"name"=>"silvano", "breed"=>"the cat kind", "id"=>4, "dob"=>"BAD VALUE"}], @gapi=#<Google::Apis::BigqueryV2::InsertAllTableDataResponse:0x007fcbce20c168 @insert_errors=[#<Google::Apis::BigqueryV2::InsertAllTableDataResponse::InsertError:0x007fcbcf01b6f0 @errors=[#<Google::Apis::BigqueryV2::ErrorProto:0x007fcbcf01adb8 @debug_info="generic::invalid_argument: Could not parse 'BAD VALUE' as a timestamp. Required format is YYYY-MM-DD HH:MM[:SS[.SSSSSS]]", @location="dob", @message="Could not parse 'BAD VALUE' as a timestamp. Required format is YYYY-MM-DD HH:MM[:SS[.SSSSSS]]", @reason="invalid">], @index=1>, #<Google::Apis::BigqueryV2::InsertAllTableDataResponse::InsertError:0x007fcbcf019670 @errors=[#<Google::Apis::BigqueryV2::ErrorProto:0x007fcbcf018f18 @debug_info="", @location="", @message="", @reason="stopped">], @index=0>], @kind="bigquery#tableDataInsertAllResponse">>
irb(main):007:0> insert_response.error_count
=> 2
irb(main):009:0> insert_response.insert_errors.first.index
=> 1
irb(main):010:0> insert_response.insert_errors.last.index
=> 0
```

**Comment by quartzmo** [comment]
However, in my manual testing, both the already-existing method `#errors_for` and the new method `#index_for` that is based on it fail to correctly return values:

```ruby
irb(main):016:0> insert_response.errors_for(rows.first)
=> []
irb(main):017:0> insert_response.errors_for(rows.last)
=> []
irb(main):018:0> insert_response.index_for(rows.first)
=> nil
irb(main):019:0> insert_response.index_for(rows.last)
=> nil
```

@zedalaye Can you give this a try and/or add an acceptance test?

**Comment by zedalaye** [comment]
I'm trying to figure out how to run acceptance tests. For now I can only run `rake test` from my fork. When I try to run `rake ci` I get this error : 

> The git source https://github.com/GoogleCloudPlatform/google-cloud-ruby.git is not yet checked out. Please run `bundle install` before trying to start your application

Of course I have run `bundle install`

**Comment by zedalaye** [comment]
Got it. I have to run `rake ci[bundleupdate]`

**Comment by zedalaye** [comment]
@quartzmo in your manual acceptance tests, shouldn't you activate skip_invalid option ?

```
irb(main):005:0> rows = [{ name: "silvano", breed: "the cat kind",      id: 4, dob: Time.now.utc }, { name: "silvano", breed: "the cat kind",      id: 4, dob: "BAD VALUE" }]
irb(main):006:0> insert_response = table.insert rows skip_invalid: true
```

**Comment by quartzmo** [comment]
@zedalaye You're right!!! I'll try again without `skip_invalid`.

You can run the acceptance tests alone with `rake acceptance`.
You can run all CI checks including acceptance tests with `rake ci:a`.

**Comment by quartzmo** [comment]
I tried again just now with ``, I still don't see the right behavior from ``:

```ruby
irb(main):006:0> result = table.insert rows, skip_invalid: true
=> #<Google::Cloud::Bigquery::InsertResponse:0x007fc3178e9e50 @rows=[{"name"=>"silvano", "breed"=>"the cat kind", "id"=>4, "dob"=>"2017-07-07 15:37:48.938740+00:00"}, {"name"=>"silvano", "breed"=>"the cat kind", "id"=>4, "dob"=>"BAD VALUE"}], @gapi=#<Google::Apis::BigqueryV2::InsertAllTableDataResponse:0x007fc3171f8e20 @insert_errors=[#<Google::Apis::BigqueryV2::InsertAllTableDataResponse::InsertError:0x007fc3171f8010 @errors=[#<Google::Apis::BigqueryV2::ErrorProto:0x007fc3178eb868 @debug_info="generic::invalid_argument: Could not parse 'BAD VALUE' as a timestamp. Required format is YYYY-MM-DD HH:MM[:SS[.SSSSSS]]", @location="dob", @message="Could not parse 'BAD VALUE' as a timestamp. Required format is YYYY-MM-DD HH:MM[:SS[.SSSSSS]]", @reason="invalid">], @index=1>], @kind="bigquery#tableDataInsertAllResponse">>
irb(main):008:0> result.error_count
=> 1
irb(main):009:0> result.insert_errors.count
=> 1
irb(main):010:0> insert_error = result.insert_error_for(rows.first)
=> nil
irb(main):011:0> insert_error = result.insert_error_for(rows.last)
=> nil
irb(main):012:0> result.insert_errors.last
=> #<Google::Cloud::Bigquery::InsertResponse::InsertError:0x007fc314ac69d8 @index=1, @row={"name"=>"silvano", "breed"=>"the cat kind", "id"=>4, "dob"=>"BAD VALUE"}, @errors=[{"debugInfo"=>"generic::invalid_argument: Could not parse 'BAD VALUE' as a timestamp. Required format is YYYY-MM-DD HH:MM[:SS[.SSSSSS]]", "location"=>"dob", "message"=>"Could not parse 'BAD VALUE' as a timestamp. Required format is YYYY-MM-DD HH:MM[:SS[.SSSSSS]]", "reason"=>"invalid"}]>
irb(main):013:0> rows.last
=> {:name=>"silvano", :breed=>"the cat kind", :id=>4, :dob=>"BAD VALUE"}
```

Curious what you will see in acceptance testing.

**Comment by quartzmo** [comment]
I think I see why now. The rows I am providing use symbol keys, but these keys are converted to strings. Because of this difference the comparison in `#insert_error_for` returns false. Since this is the original implementation, I don't think addressing it is in the scope of this PR. However, a passing acceptance test would still be nice. (You probably want to use string keys in your test.)

**Comment by zedalaye** [comment]
I found the root of the problem:

```ruby
irb(main):010:0> insert_error = result.insert_error_for(rows.first)
=> nil
irb(main):011:0> insert_error = result.insert_error_for(rows.last)
=> nil
```

That fails because you first have to Convert.to_json_row(rows.first) to be able to use it as key index for theses methods. Look at table#insert, you'll see that it convert rows before sending data to BigQuery and use them as this in the response data.

Given this acceptance test data:
```ruby
  let(:invalid_rows) do
    [
      { name: "silvano", breed: "the cat kind",      id: 4, dob: Time.now.utc },
      { name: nil,       breed: "golden retriever?", id: 5, dob: Time.now.utc },
      { name: "stephen", breed: "idkanycatbreeds",   id: 6, dob: Time.now.utc }
    ]
  end
````

I finally succeeded in writing this acceptance test that works:
```ruby
  it "insert skip invalid rows and return insert errors" do
    # data = table.data
    insert_response = table.insert invalid_rows, skip_invalid: true
    insert_response.wont_be :success?
    insert_response.insert_count.must_equal 2

    insert_response.insert_errors.wont_be :empty?
    insert_response.insert_errors.count.must_equal 1
    insert_response.insert_errors.first.class.must_equal Google::Cloud::Bigquery::InsertResponse::InsertError
    insert_response.insert_errors.first.index.must_equal 1

    # In the context of this test, the original row cannot be compared with InsertResponse row because
    # rows are converted to "BigQuery JSON rows" early in table.insert: key symbols are turned into strings and
    # dates/times are formated as timestamp strings with milliseconds.
    # To be able to test InsertResponse#{insert_error_for, errrors_for, index_for} we need a "BigQuery JSON row"
    # instead of using insert_response.insert_error.first we make one by converting our orginal "invalid_row"
    bigquery_row = Google::Cloud::Bigquery::Convert.to_json_row(invalid_rows[insert_response.insert_errors.first.index])

    insert_response.insert_errors.first.row.must_equal bigquery_row

    insert_response.error_rows.wont_be :empty?
    insert_response.error_rows.count.must_equal 1
    insert_response.error_rows.first.must_equal bigquery_row

    insert_response.insert_error_for(bigquery_row).index.must_equal insert_response.insert_errors.first.index
    insert_response.errors_for(bigquery_row).wont_be :empty?
    insert_response.index_for(bigquery_row).must_equal 1
  end
```

Could you please reword my comments to something clearer/better ?

**Comment by quartzmo** [comment]
I understand your comment perfectly; for me, no change is needed.

However, the problem I see now is that `Google::Cloud::Bigquery::Convert` is not part of the documented API (marked with `@private`).

Maybe the library can automatically apply the conversion for the user within `#insert_error_for`?

**Comment by zedalaye** [comment]
> Maybe the library can automatically apply the conversion for the user within #insert_error_for?

That was very easy to do finally. Thank you for suggesting that change.

**Comment by quartzmo** [comment]
Great! Can you see if you can appease @googlebot (see the suggestions in [the comment above](https://github.com/GoogleCloudPlatform/google-cloud-ruby/pull/1579#issuecomment-313724351)).

**Comment by zedalaye** [comment]
I signed a company wide CLA (got the email and DocuSigned it) but I cannot find it in https://cla.developers.google.com/clas in "Manage Agreements". I don't know what to do more.

**Comment by quartzmo** [comment]
> maybe you used a different email address in the git commits than was used to sign the CLA?

Do you think maybe this is the case? So sorry for the trouble.

File: google-cloud-bigquery/lib/google/cloud/bigquery/insert_response.rb [hunk 1]
```diff
@@ -53,12 +53,22 @@ def error_rows
           end
         end
 
+        def insert_error_for row
+          insert_errors.detect { |e| e.row == Convert.to_json_row(row) }
```
Comment (blowmage) [inline google-cloud-bigquery/lib/google/cloud/bigquery/insert_response.rb]: Better to convert the row once, and not for every iteration.

```ruby
json_row = Convert.to_json_row row
insert_errors.detect { |e| e.row == json_row }
```

Comment (quartzmo) [inline google-cloud-bigquery/lib/google/cloud/bigquery/insert_response.rb]: Yes, agreed. Good catch @blowmage.

**Comment by zedalaye** [comment]
@blowmage I pushed the change you suggested. Thank you !

**Comment by blowmage** [comment]
Thanks man. This looks good to me. However, we are unable to accept this PR until you get the CLA thing figured out.

**Comment by blowmage** [comment]
Ah, you might want to squash all commits into one. It looks like you have a merge commit that was authored by a different user. If you squash all the commits and force push it should fix the CLA check.

**Comment by quartzmo** [comment]
>  It looks like you have a merge commit that was authored by a different user.

Thank you for noticing this @blowmage!

**Comment by zedalaye** [comment]
Yeah. Squashing commits have solved the GoogleBot CLA issue ! Thank you @blowmage

**Comment by quartzmo** [comment]
@zedalaye Thank you for this contribution! We will release a new version including your change very soon.

**Comment by zedalaye** [comment]
Thank you @quartzmo ! Thank you two for your advices and your support.

**Comment by quartzmo** [comment]
@zedalaye This was published in [google-cloud-bigquery 0.27.1](https://rubygems.org/gems/google-cloud-bigquery/versions/0.27.1) and [google-cloud 0.35.0](https://rubygems.org/gems/google-cloud/versions/0.35.0).

**Pull request merged**